### PR TITLE
Scientific notation numbers with decimal separators support

### LINKED ISSUE #39139: JsonElement.WriteAsValue does not support scientific notation with decimal separator
I encountered the issue where following code throws:
```c#
var test = JsonDocument.Parse("{ \"test\" : 6.5e-5 }").RootElement;

using(var stream = new MemoryStream())
{
       var writer = new Utf8JsonWriter(stream);
       //throws System.ArgumentException: 'e' is an invalid end of a number. Expected a delimiter.
       test.WriteAsValue(writer);       
}
```

Tested on .net core 3 preview5 and preview6.

### PR DESCRIPTION
**Pull request by vicrdguez**
Writing Scientific notation numbers with decimal separators and exponent throwed an exception.
The decimal separator and the exponent separator were condition exclusive in the number validation

Fixes #39139.

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 1]
```diff
@@ -172,21 +172,38 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
...
+                while (i < utf8FormattedNumber.Length && JsonHelpers.IsDigit(utf8FormattedNumber[i]))
+                {
+                    i++;
+                }
             }
-            else if (val == 'e' || val == 'E')
+
+            if (utf8FormattedNumber.Length <= i)
+            {
+                throw new ArgumentException(SR.RequiredDigitNotFoundEndOfData, nameof(utf8FormattedNumber));
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: This should only throw for greater-than (or, with this presentation, less-than); as equal is legal (per the next if).

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 2]
```diff
@@ -172,21 +172,38 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
...
+            }
+
+            val = utf8FormattedNumber[i];
+
+            if (val == 'e' || val == 'E')
             {
                 i++;
 
-                if (i >= utf8FormattedNumber.Length)
+                if (utf8FormattedNumber.Length <= i)
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Why are you reversing the comparison?

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs [hunk 3]
```diff
@@ -3747,6 +3747,13 @@ public void WriteNumbers(bool formatted, bool skipValidation, string keyString)
             doubles[2] = double.MinValue;
             doubles[3] = 12.345e1;
             doubles[4] = -123.45e1;
+            doubles[5] = 6.022e-23;
+            doubles[6] = -6.022e-23;
+            doubles[7] = 1e200;
+            doubles[8] = -1e200;
+            doubles[9] = 1e-200;
+            doubles[10] = -1e-200;
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: While these may be reasonable tests, these aren't testing the code change.

To test the code change you'd need to add a test like

```C#
[Theory]
[InlineData("6.022e-23", false)]
[InlineData("6.022e-23", true)]
[InlineData("6.022e23", false)]
[InlineData("6.022e23", true)]
[InlineData("6.022e+23", false)]
[InlineData("6.022e+23", true)]
[InlineData("-6.022e-23", false)]
[InlineData("-6.022e-23", true)]
[InlineData("-6.022e23", false)]
[InlineData("-6.022e23", true)]
[InlineData("-6.022e+23", false)]
[InlineData("-6.022e+23", true)]
public static void WriteNumberDecimalScientific(string value, bool indented)
{
    WriteSimpleValue(indented, value);
}
```

In JsonElementWriteTests.cs, around line 39 (between WriteNumberScientific and WriteNumberOverprecise).

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Actually, it'd be better move the next if (`i == utf8FormattedNumber.Length`) to inside the `if (val == '.')` block, after the while.  (So this check looks like the rest of the "finished too early" checks)

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs (hunk 2 above)
Comment (vicrdguez) [reply src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: I saw some other comparison in that order, just to have all the code looking the same

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs (hunk 3 above)
Comment (vicrdguez) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: roger that I didn't get it right

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 4]
```diff
@@ -172,21 +172,38 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
                 return;
             }
 
+            //The non digit character inside the number
```
Comment (danmosemsft) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Nit, space after //

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs (hunk 1 above)
Comment (danmosemsft) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Should these be ThrowHelper like elsewhere in the file?

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 5]
```diff
@@ -203,11 +220,9 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
                     nameof(utf8FormattedNumber));
             }
 
-            if (i >= utf8FormattedNumber.Length)
+            if (utf8FormattedNumber.Length <= i)
             {
-                throw new ArgumentException(
-                    SR.RequiredDigitNotFoundEndOfData,
-                    nameof(utf8FormattedNumber));
+                throw new ArgumentException(SR.RequiredDigitNotFoundEndOfData, nameof(utf8FormattedNumber));
```
Comment (danmosemsft) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Do the tests hit every one of these throws?

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 6]
```diff
@@ -140,6 +140,7 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
             // because it doesn't need to deal with "NeedsMoreData", or remembering the format.
             if (utf8FormattedNumber.IsEmpty)
             {
+                //ThrowHelper.ThrowArgumentException(nameof(utf8FormattedNumber), utf8FormattedNumber);
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Please don't add commented out code.  If there's a ThrowHelper for this already, use it. If not, just delete the comment and be a simple throw.

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs (hunk 4 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: @vicrdguez Our comment style is to have a space after the `//` and before the words.  If you feel like this comment adds value, please insert the space.

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs (hunk 3 above)
Comment (bartonjs) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: Please remove these extra inputs, as this is an unrelated change.  (These numbers were parsed by the compiler, not the JSON writer, they aren't evaluated, just written down via Utf8Formatter... because they can't be "invalid")

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 7]
```diff
@@ -56,19 +84,20 @@ public static void WriteNumberOverprecise(bool indented)
...
 
         [Theory]
-        [InlineData(false)]
-        [InlineData(true)]
-        public static void WriteNumberTooLargeScientific(bool indented)
+        [InlineData("1e400", false)]
+        [InlineData("1e400", true)]
+        [InlineData("-1e400", false)]
+        [InlineData("-1e400", true)]
+        public static void WriteNumberTooLargeScientific(string value, bool indented)
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: Please revert the changes to this particular test.  The formatting variation is already covered by the above two tests, and these changes have made the comments explaining the value of this test case be awkwardly not adjacent to the value that they're explaining.  Also, now there are values instead of a single value.

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs (hunk 5 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Currently hitting the throws requires either a bug in this method (how we got here :smile:) or a JsonElement having gotten into an invalid/corrupted state. (This method is here to guard against that corruption and to serve as validation for a future addition to the writer where it allows you to write the numeric string directly)

Having written both this method and JsonElement, it's probably most efficient for me to write the tests to get the coverage on these throws... after we get this PR in so we get the broken success scenario fixed.

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 8]
```diff
@@ -172,21 +173,39 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
...
+                    i++;
+                }
+
+                if (utf8FormattedNumber.Length < i)
+                {
+                    throw new ArgumentException(SR.RequiredDigitNotFoundEndOfData, nameof(utf8FormattedNumber));
+                }
+            }
+
+
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: Please remove the second of the two blank lines.  (Or, if you prefer, the first... not like I can tell :))

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs]: ThrowHelper is usually used to change the inlining cost/value of a method.  If there's not already a ThrowHelper for this, it's probably not worth adding one.  I don't think there is, because this is "reader logic", so it'd be in the reader's ThrowHelper, which requires passing in a ref reader to explain how you go to the bad state.

In general, we prefer `throw new` over ThrowHelper, doing the latter only where needed/justified.

Everything else in this file is a very small method that is trying to be an inline candidate.  This is probably too big to inline even with ThrowHelper.

**Comment by ericstj** [comment]
Fixes #39139

**Comment by bartonjs** [comment]
The JSON tests have already reported in on the Windows x86 Release run (no failures, but 5 other library answers haven't posted yet), and this change is likely going to cause another merge conflict for another JSON change outstanding; so using my judgment to go ahead and merge.
